neonvm/pkg/ipam/types.go

```diff
@@ -0,0 +1,41 @@
+package ipam
+
+import (
+	"net"
+
+	cnitypes "github.com/containernetworking/cni/pkg/types"
+)
+
+type temporaryError struct {
+	error
+}
+
+func (t *temporaryError) Temporary() bool {
+	return true
+}
+
+type RangeConfiguration struct {
+	OmitRanges []string `json:"exclude,omitempty"`
+	Range      string   `json:"range"`
+	RangeStart net.IP   `json:"range_start,omitempty"`
+	RangeEnd   net.IP   `json:"range_end,omitempty"`
+}
+
+type Nad struct {
+	IPAM *IPAMConfig `json:"ipam"`
+}
+
+// IPAMConfig describes the expected json configuration for this plugin
+type IPAMConfig struct {
+	Routes           []*cnitypes.Route    `json:"routes"`
+	IPRanges         []RangeConfiguration `json:"ipRanges"`
+	OmitRanges       []string             `json:"exclude,omitempty"`
+	DNS              cnitypes.DNS         `json:"dns"`
+	Range            string               `json:"range"`
+	RangeStart       net.IP               `json:"range_start,omitempty"`
+	RangeEnd         net.IP               `json:"range_end,omitempty"`
+	GatewayStr       string               `json:"gateway"`
+	Gateway          net.IP
+	NetworkNamespace string
```

Comment:
These json fields seem... strange. Is that intended?

Reply:
right now - no, but maybe in the future we would like to implement some routing between overlay and pod networks.

Reply:
I don't quite follow. As written, a value like:

```go
IPAMConfig{
	// ...
	GatewayStr:       "127.0.0.1",
	Gateway:          net.ParseIP("127.0.0.2"),
	NetworkNamespace: "foo",
	NetworkName:      "bar",
}
```

will be serialized like:

```json
{
    ...
    "gateway":          "127.0.0.1",
    "Gateway":          "127.0.0.2",
    "NetworkNamespace": "foo",
    "network_name":     "bar",
}
```

If that's not intended, it should be fixed.

(see also: `net.IP` [produces a JSON string when marshalled](https://go.dev/play/p/Tei2t9wl9hJ))

Reply:
added to follow ups list

Reply:
Where are these types used? If the JSON serialization is stored somewhere, this cannot be left for a follow-up because this will cause a breaking change.

And even if it's not stored, this is just sloppy. Please reconsider.

Reply:
`Gateway` filled by LoadFromNad function from `GatewayStr`. There are no place where you can specify `Gateway` directly

https://github.com/neondatabase/autoscaling/blob/networking-k8s-v1.25/neonvm/pkg/ipam/ipam.go#L159-L166

Reply:
I must've missed it in my initial review — AFAICT having separate `Gateway` vs `GatewayStr` is unnecessary. You can directly unmarshal JSON into a `net.IP`: https://go.dev/play/p/voZcypEqSAW

Reply:
removed `gateway*` at all